Scala 3 support

jjudd opened the pull request:
Add Scala 3 support, update to 2.13, and update Zinc

This is a rather large commit with lots of changes. The end goal of the
commit is to support Scala 3. That goal has been met, but not extensively
tested in a production codebase.

Part of this effort included upgrading the Scala 2 version in these rules
to Scala 2.13 in order to test Scala 2.13's forwards compatibility with
Scala 3.

Zinc was also upgraded, so it supports Scala 3.

Scalafmt was also upgraded, so it supports scala 3.

Commit:
Add Intellij support

Commit:
Add Scala 3 support, update to 2.13, and update Zinc

This is a rather large commit with lots of changes. The end goal of the
commit is to support Scala 3. That goal has been met, but not extensively
tested in a production codebase.

Part of this effort included upgrading the Scala 2 version in these rules
to Scala 2.13 in order to test Scala 2.13's forwards compatibility with
Scala 3.

Zinc was also upgraded, so it supports Scala 3.

Scalafmt was also upgraded, so it supports scala 3.

jjudd commented on .bazelrc_shared:
```diff
@@ -6,24 +6,27 @@ build --javacopt="-source 8 -target 8"
 build --strategy=ScalaCompile=worker
 build --worker_max_instances=4
-build --worker_sandboxing
+#build --worker_sandboxing
```
TODO: Re-enable this and make sure it works

jjudd replied:
Fixed by undoing [this commit](https://github.com/lucidsoftware/rules_scala/commit/037d7440cf3b192b9979ff2f2c1b2c674b433afc)

I think sandboxing got better in 4 and that is what broke this

jjudd commented on rules/scala/workspace.bzl:
```diff
@@ -11,24 +12,27 @@ filegroup(
+        "org.scala-sbt:zinc-compile-core_2.13:1.5.7",
+        "org.scala-sbt:zinc-core_2.13:1.5.7",
+        "org.scala-sbt:zinc-persist_2.13:1.5.7",
+        maven.artifact("org.scala-sbt", "compiler-bridge_2.13", "1.5.7", neverlink = True),
```
Add a comment why this is neverlink

jjudd replied:
Done.

jjudd commented on rules/scala/workspace_2_12.bzl:
```diff
@@ -0,0 +1,23 @@
+        "org.scala-lang:scala-compiler:2.12.14",
+        "org.scala-lang:scala-library:2.12.14",
+        "org.scala-lang:scala-reflect:2.12.14",
+        maven.artifact("org.scala-sbt", "compiler-bridge_2.12", "1.5.7", neverlink = True),
```
TODO add comment here as well

jjudd replied:
Done

jjudd commented on rules/scala/workspace_3.bzl:
```diff
@@ -0,0 +1,23 @@
+        "org.scala-lang:scala3-compiler_3:3.0.2",
+        "org.scala-lang:scala3-library_3:3.0.2",
+        "org.scala-sbt:compiler-interface:1.5.7",
+        maven.artifact("org.scala-lang", "scala3-sbt-bridge", "3.0.2", neverlink = True),
```
TODO: add comment here

jjudd replied:
Done

jjudd commented on src/main/scala/BUILD:
```diff
@@ -7,53 +7,99 @@ load(
+    compiler_bridge = "@annex_3//:org_scala_lang_scala3_sbt_bridge_3_0_2",
+    compiler_classpath = compiler_classpath_3_0_2,
+    runtime_classpath = runtime_classpath_3_0_2,
+    version = "3.0.1",
```
TODO: version is wrong

jjudd replied:
Done

jjudd commented on src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala:
```diff
@@ -1,30 +1,74 @@
+  override val libraryJars: Array[File] = allJars
+    .filter(f => new URLClassLoader(Array(f.toURI.toURL)).findResource("library.properties") != null)
+
+  override val otherJars: Array[File] = allJars.diff(compilerJars ++ libraryJars)
```
TODO: this seems wrong

SrodriguezO replied:
Looking ScalaInstance.scala linked above, this does seem right :thinking: I wonder if it's legacy? afaict, all the constructors just set `compilerJars` to be `allJars`, and they do set `otherJars` to be only jars that are not `compilerJars` or `libraryJars`, so afaict, `otherJars` is always empty

jjudd replied:
Thanks for looking into this. Glad to know I got it right :D

jjudd commented on tests/compile/logger/Example.scala:
```diff
@@ -10,7 +10,7 @@ import xsbti.Severity
     def main(args: Array[String]) {
         val logger = new AnnexLogger(LogLevel.Info)
-        val reporter = new LoggedReporter(logger)
+        val reporter = new LoggedReporter(logger, "2.13.6")
```
Why does this have a version string?

jjudd commented on tests/compile/srcjar/test:
```diff
@@ -1,8 +1,8 @@
 #!/bin/bash -e
-. "$(dirname "$0")"/../../common.sh
+#. "$(dirname "$0")"/../../common.sh
```
TODO: Found a test that's still commented

jjudd replied:
Done

coreywoodfield commented on rules/scala_proto/private/ScalaProtoWorker.scala:
```diff
@@ -40,12 +40,21 @@ object ScalaProtoWorker extends WorkerMain[Unit] {
     Files.createDirectories(scalaOut)
 
-    val params = s"--scala_out=$scalaOut" :: sources.map(_.getPath)
+    val params = List(s"--scala_out=$scalaOut")
```
The protobuf rules have been improved [in the upstream repo](https://github.com/higherkindness/rules_scala/commit/e9f7843366472a36c48b793f0124f93721b7b081). I wonder if just pulling in the upstream changes would be sufficient to solve the issues you were having with protobuf.

SrodriguezO replied:
Trying to undo these changes and pulling in the upstream files might be worth trying. 

Alternatively, we can keep this and then replace after this PR lands if we wanna break up the change, but then these files need a review. I'm not as familiar with the Proto part of the rules; @coreywoodfield would you mind reviewing the files under `rules/scala_proto` if we go this route?

jjudd replied:
In in favor of landing this change then getting the lucid-master branch and upstream synced up, which would involve dealing with the proto stuff.

coreywoodfield replied:
`rules/scala_proto` changes look good to me

coreywoodfield commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:
```diff
@@ -12,48 +12,34 @@ import java.io.{File, PrintStream, PrintWriter}
  * Zinc has two caches:
- *  1. a ClassLoaderCache which is a soft reference cache for classloaders of Scala compilers.
- *  2. a CompilerCache which is a hard reference cache for (I think) Scala compiler instances.
+ *   1. a ClassLoaderCache which is a soft reference cache for classloaders of Scala compilers. 2. a CompilerCache which
```
Uhh this formatting change is weird

jjudd replied:
Agreed. Scalafmt did it, so I undid it.

coreywoodfield commented on tests/plugins/classpath/test:
```diff
@@ -1,9 +1,11 @@
-#!/bin/bash -e
+# !/bin/bash -e
```
Why was this space added?

jjudd replied:
My mistake. Fixed it.

coreywoodfield commented on tests/plugins/kind-projector/KindProjector.scala:
```diff
@@ -1,13 +1,12 @@
 package anx
```
This file doesn't seem to be using kind-projector anymore, which seems to defeat its purpose

jjudd replied:
Good catch. I was playing around with fixing this test and left it in a passing, but useless state. This is fixed.

SrodriguezO replied:
I'm not familiar w/ this, so I'll let @coreywoodfield resolve this comment :upside_down_face:

SrodriguezO replied:
discussed offline; @coreywoodfield said it looks good

coreywoodfield commented on tests/test-frameworks/mixed/test:
```diff
@@ -1,9 +1,14 @@
-bazel test :mixed_2_11 --test_arg=--verbosity=MEDIUM | grep -v 'org.scalacheck.ScalaCheckFramework: 1 tests'
+function test ()  {
+  out=$(bazel test ":mixed_$1")
+  echo "$out" | grep 'should pop values in last-in-first-out order'
```
Is this checking for the line that was removed from the previous file? This should perhaps also be changed

SrodriguezO replied:
mostly curious, but why did scalatest.scala need to change? not opposed to it, but afaict we just swapped a dummy test for another :shrug:

jjudd replied:
I upgrade ScalaTest as part of this PR and the old test no longer compiled with the new version. So I created a new dummy test.

jjudd replied:
Good catch on that grep. Putting it inside of a bash function broke the exit on error, making this test pass every time. After re-reading the original test, it's also not a good idea to make this a bash function because the code isn't duplicated.

Everything should be fixed now and the test should once again be useful.

SrodriguezO commented on rules/private/phases/phase_javainfo.bzl:
```diff
@@ -43,7 +43,9 @@ def phase_javainfo(ctx, g):
         java_info = JavaInfo(
-            compile_jar = compile_jar,
+            # TODO: why do ijars break Scala 3?
+            # compile_jar = compile_jar,
```
nit: might be good to also add some details on the anticipated impact of disabling this here (increased cache bust impact on non-interface changes)

SrodriguezO replied:
If we're disabling ijars, let's comment out
```
compile_jar = java_common.run_ijar(
    ctx.actions,
    jar = ctx.outputs.jar,
    target_label = ctx.label,
    java_toolchain = find_java_toolchain(ctx, ctx.attr._java_toolchain),
)
```
on lines 30-35 and replace with `compile_jar = ctx.outputs.jar` (along with the comments on why it's disabled and impact).

Might as well skip that work if we're not using it.

SrodriguezO replied:
We should also do some benchmarks to ensure the impact form this won't be catastrophic; I can get those going.

jjudd replied:
I did some more digging into this. The tasty files are preserved. That change was made in Bazel. I'm beginning to wonder if the classfile format changed between Scala 2 and 3, such that ijar doesn't work on Scala 3 stuff. My guess is we'll need to dig into that stuff to figure this out. 

Some links for posterity: 
https://github.com/bazelbuild/bazel/blob/master/third_party/ijar/classfile.cc#L1507
https://github.com/bazelbuild/bazel/blob/master/third_party/ijar/ijar.cc#L40
https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

jjudd replied:
Made the changes you suggested. I recommend we merge this change without the benchmarks, so this repo supports Scala 3. Then we can wait to use it internally until we have said benchmarks.

SrodriguezO replied:
Would it be hard to make this configurable? Then we can easily toggle in on/off if it proves problematic. Granted we'd also be toggling scala3 support on/off, but that sounds reasonable to me

SrodriguezO commented on src/main/scala/BUILD:
```diff
@@ -7,53 +7,99 @@ load(
+# Adding this, so we make sure to have a Scala library in the
+# IntelliJ libraries, so we can get a Scala SDK on sync.
+scala_library(
+    name = "scala-sdk",
```
To confirm, this won't be built into any downstream jars, correct? Otherwise, might the `@annex_3` deps cause dependency conflicts?

jjudd replied:
Correct. As far as I know this just pulls the library into the workspace for IntelliJ because we don't depend on it anywhere.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/test/TestDiscovery.scala:
```diff
@@ -29,7 +29,7 @@ class TestDiscovery(framework: Framework) {
 
   private[this] def discover(definitions: Seq[Definition]) =
-    Discovery(subclassPrints.map(_.superclassName)(breakOut), annotatedPrints.map(_.annotationName)(breakOut))(
+    Discovery(subclassPrints.map(_.superclassName).to(Set), annotatedPrints.map(_.annotationName).to(Set))(
```
These are already sets; the `.to(Set)` is redundant

jjudd replied:
Ah. Nice catch. I read this line `val subclassPrints = mutable.ArrayBuffer.empty[TestSubclassFingerprint]` and missed this line `private[this] val (annotatedPrints, subclassPrints) = {`

jjudd replied:
Fixed

SrodriguezO commented on tests/plugins/kind-projector/BUILD:
```diff
@@ -4,31 +4,32 @@ scala_library(
     runtime_classpath = [
-        "@scala_library_2_12_6//jar",
+        "@annex//:org_scala_lang_scala_library_2_13_6",
+        "@org_typelevel_kind_projector_2_13//jar",
```
just curious - why do we need the kind_projector jar on the runtime classpath now?

jjudd replied:
We do not. I made this test not use the kind_projector by accident and this was part of that accident. I fixed things now.

SrodriguezO commented on tests/WORKSPACE:
```diff
@@ -142,18 +162,27 @@ load("@annex_test//:defs.bzl", annex_test_pinned_maven_install = "pinned_maven_i
+
+annex_test_2_12_pinned_maven_install()
+
 jvm_maven_import_external(
```
Can `kind_projector` and `scalacheck` come in through `rules_jvm_external`? If so, I think we should just move them there too

jjudd replied:
Yep. Done.

SrodriguezO commented on tests/scala/BUILD:
```diff
@@ -6,164 +6,135 @@ load(
     visibility = ["//visibility:public"],
 )
 
-# zinc + typelevel 2.12.1
```
Typelevel Scala's been [archived](https://github.com/typelevel/scala) for a couple years, so dropping support is probably fine at some point. That said, I don't see how its tests would interfere with Scala 3 compatibility, so imo we shouldn't remove them as part of this change

jjudd replied:
As part of this PR I updated to Scala 2.13 & 2.12 from 2.12 & 2.11. Typelevel does not have an up to date version of 2.12 or 2.13, so I removed the Typelevel Scala

SrodriguezO commented on tests/scala/scala2-12/BUILD:
```diff
@@ -3,6 +3,6 @@ load("@rules_scala_annex//rules:scala.bzl", "scala_library")
 scala_library(
     name = "example",
     srcs = ["Example.scala"],
-    scala = "//scala:zinc_typelevel_2_12_1",
```
same comment here RE removing Typelevel Scala tests

jjudd replied:
Same comment as above. Without an up to date 2.12 or 2.13 version of Scala, there is no reason to continue to support this. A Scala version upgrade seems like as good a time as any to remove it.

SrodriguezO commented on tests/scala/scala3/BUILD:
```diff
@@ -0,0 +1,43 @@
+    tags = ["manual"],
+)
+
+scala_library(
```
this one's missing the `scala` attribute

jjudd replied:
That was intentional. I wanted a Scala 2 library in the dependency tree. I'll make it explicit, so it's less confusing.

SrodriguezO commented on tests/test-frameworks/isolation/BUILD:
```diff
@@ -3,106 +3,112 @@ load("@rules_scala_annex//rules:scala.bzl", "scala_library", "scala_test")
-        "@annex_test//:org_specs2_specs2_common_2_11",
-        "@annex_test//:org_specs2_specs2_core_2_11",
-        "@annex_test//:org_specs2_specs2_matcher_2_11",
+        ":global-2-12",
```
this was previously depending on a library w/ a different Scala version - shouldn't we preserve that behavior on the test?

SrodriguezO replied:
That's kind of confusing though - these tests aren't for cross-compile, they're for test isolation levels :thinking: I don't think deps w/ different Scala versions would affect that

jjudd replied:
Agreed. This is a bug that we didn't catch because 2.12 and 2.11 aren't as different as 2.12 and 2.13. Changing this to the 2.13 global causes compilation to fail. So, I'm going to consider this a bug fix and leave the change.

SrodriguezO commented on tests/compat/compat.iml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
We should add `*.iml` to the `.gitignore`

jjudd replied:
Good catch. Done.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:
```diff
@@ -211,19 +200,19 @@ object ZincRunner extends WorkerMain[Namespace] {
         .withClassLoaderCache(classloaderCache)
       lastCompiler = scalaCompiler
-      ZincUtil.compilers(scalaInstance, ClasspathOptionsUtil.boot, None, scalaCompiler)
+      ZincUtil.compilers(scalaInstance, ClasspathOptionsUtil.auto, None, scalaCompiler)
```
This change adds the following two conditions on the returned Compilers instance:
```
2. the Compiler JAR is present in the classpath;
3. the extra JARs present in the Scala instance are added to the classpath.
```

Why is this change needed / what is its impact?

jjudd replied:
I do not remember why I made this change and tests pass without it, so I'm going to go back to boot. My guess is I made this change while dealing with classpath stuff and forgot to undo it. Or I made it for some important reason that we'll quickly find out.

jjudd replied:
I'd leave it, but I want to minimize classpath leakage stuff and your point 3 above concerns me.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -2,47 +2,58 @@ package higherkindness.rules_scala
+import sbt.internal.inc.Schema.Type.{Projection, Structure}
+import sbt.internal.inc.{APIs, Analysis, PlainVirtualFile, PlainVirtualFileConverter, Relations, Schema, SourceInfos, Stamp => StampImpl, Stamper, Stamps}
+import sbt.internal.inc.Schema.{AnalyzedClass, Annotation, ClassDependencies, ClassLike, Companions, NameHash, Type, TypeParameter, UsedName, UsedNames, Values}
+import sbt.internal.shaded.com.google.protobuf.{GeneratedMessage, GeneratedMessageV3, Message}
```
nit: you can remove `GeneratedMessage` and `Message` now

jjudd replied:
Done.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -107,14 +122,16 @@ class AnxAnalysisStore(files: AnalysisFiles, analyses: AnxAnalyses) extends Anal
-      id = value.id,
-      prefix = value.prefix,
-    )
+    // I copied this over from the old code, but I'm pretty sure this isn't doing anything useful?
```
~~from what old code?~~ Oh.. you just changed the `value.copy` to a new instantiation, gotcha. Why though?

jjudd replied:
Can't use copy because it is a Java object rather than a Scala object. I don't think it's doing anything useful, but I want to remove it in another, smaller change, so if it breaks something it is easier to figure out that removing this broke it.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -107,14 +122,16 @@ class AnxAnalysisStore(files: AnalysisFiles, analyses: AnxAnalyses) extends Anal
   private[this] val writer = new ProtobufWriters(mappers.getWriteMapper)
 
   def sortProjection(value: Projection): Projection = {
-    value.copy(
```
:thinking: was this even doing anything?

jjudd replied:
I don't believe so, no.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -124,166 +141,340 @@ class AnxAnalyses(format: AnxAnalysisStore.Format) {
+      .newBuilder(annotation)
+      .clearArguments()
+      .addAllArguments(sortedArguments)
+      .build()
```
these sort changes lgtm, but I'm curious - what was wrong with the previous object copy approach?

SrodriguezO replied:
The copy had the advantage that you everything else stayed the same. In this case you have to make sure you're setting/unsetting the right things (e.g. you had to clearArguments before adding the sorted arguments here)

jjudd replied:
The previous protobuf stuff was a Scala library. Their replacement is Java. So copy is not an option because it's not a Scala case class, rather it is a Java class.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -124,166 +141,340 @@ class AnxAnalyses(format: AnxAnalysisStore.Format) {
+  def updateApisWithDefaultCompilationTimestamp(apis: Schema.APIs): Schema.APIs = {
+    val internal = apis.getInternalMap.asScala.iterator
+      .map { case (s: String, analyzedClass: AnalyzedClass) =>
+        val foo = updateAnalyzedClassWithDefaultCompilationTimestamp(analyzedClass)
```
nit: skip the foo
```
(s, updateAnalyzedClassWithDefaultCompilationTimestamp(analyzedClass))
```

jjudd replied:
Thanks. Nice catch. Very likely this is leftover from me debugging something.

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -124,166 +141,340 @@ class AnxAnalyses(format: AnxAnalysisStore.Format) {
+      .clearClasses()
+      .putAllClasses(updatedClasses)
+      .build()
+//    _.update(
```
nit: leave a comment explaining that these commented out lines are just the old logic or remove them altogether

jjudd replied:
Nice catch

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/zinc/compile/AnalysisStore.scala:
```diff
@@ -124,166 +141,340 @@ class AnxAnalyses(format: AnxAnalysisStore.Format) {
+      .asJava
+    // The original code mutated this twice and I'm not totally sure why. Not sure if it is a bug or if the
+    // scalapb version the mutations were applied sequentially. I've tried to write the code to apply the
+    // mutations sequentially, assuming that is how that code works.
```
This seems correct insofar as it's doing the same thing the previous logic was doing. That said, I don't fully understand what the previous logic was doing.. I dislike this method lol - hopefully we can indeed remove it now that zinc-532 is closed.. we can try that on a small follow-up PR.

jjudd replied:
Agree

SrodriguezO commented on src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala:
```diff
@@ -1,30 +1,74 @@
+      // Problem is, doing this creates a potential leak: the parent classloader will be checked
+      // for all classes used during the compilation. Meaning, any classes used in the
+      // ZincRunner could end up leaking out into the Scala 3 compiler. So far this hasn't
+      // caused any issues, but maybe it will in the future.
```
Would it be safer to go the other way around and make this classloader the parent for the ZincRunner's classloader? It seems classes leaking into the Scala compiler would be more problematic than into the ZincRunner

jjudd replied:
The ZincRunner's classloader is the classloader that happens when we launch the JVM. I am ignorant of whether you can/how you can control that classloader or do what you are suggesting. Not saying it's not possible, just that I don't know how. Essentially, I think the answer to your question is yes, but I don't know how to do it or whether it is possible. Especially considering this classloader is not created until well after the program has started executing.

I do think revisiting this in the future would be a good idea, but so far I haven't run into any issues with the leak.

SrodriguezO reviewed (changes requested):
Overall looks great :) I left a bunch of comments - some nits, some just seeking clarification. There's also some other comments left by other reviewers and a few tests pending re-enable.

*Please add new commits and do not squash* :pray: it's a gigantic PR and going through all the files took a while :sweat:

jjudd commented:
Changes requested have been made @SrodriguezO @coreywoodfield. All my changes are in new commits. I force pushed because I wanted to pick up the latest commits from lucid-master.

jjudd commented:
Just pulled in the GitHub actions build change, so we can run builds again. Tests are passing locally, but I need to get them passing GitHub CI.

SrodriguezO commented on rules/private/phases/phase_javainfo.bzl:
```diff
@@ -27,12 +27,19 @@ def phase_javainfo(ctx, g):
+        #     target_label = ctx.label,
+        #     java_toolchain = find_java_toolchain(ctx, ctx.attr._java_toolchain),
+        # )
+        compile_jar = ctx.outputs.jar
```
commented on the other thread, but adding here too - I think we should make this configurable

jjudd replied:
Done

SrodriguezO commented on tests/test-frameworks/mixed/test:
```diff
@@ -1,14 +1,10 @@
-test "2_12"
+bazel test :mixed_2_12 | grep 'Dummy.bar: OK, proved property.'
+bazel test :mixed_2_12 --test_arg=--verbosity=HIGH | grep 'org.scalacheck.ScalaCheckFramework: 1 tests'
+bazel test :mixed_2_12 --test_arg=--verbosity=MEDIUM | grep -v 'org.scalacheck.ScalaCheckFramework: 1 tests'
```
why do we only do the first two tests w/ 2.13 and the last three w/ 2.12 now instead of all 5 w/ each?

jjudd replied:
Because that's how this was originally. I made a mistake when I moved them all into the function. It looks like this https://github.com/lucidsoftware/rules_scala/blob/lucid-master/tests/test-frameworks/mixed/test on the master branch.

SrodriguezO commented on third_party/bazel/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/jarhelper/JarHelper.java:
```diff
@@ -53,8 +66,6 @@
   public static final Attributes.Name INJECTING_RULE_KIND =
       new Attributes.Name("Injecting-Rule-Kind");
 
-  public static final long DOS_EPOCH_IN_JAVA_TIME = 315561600000L;
```
why remove this?

jjudd replied:
I updated this file to the most recent version upstream. And this was removed upstream.

SrodriguezO reviewed (changes requested):
Pretty much there - just a couple last clarifications, and making the ijars/scala3 tradeoff configurable. It could be added to the `ScalaConfiguration` provider and passed in through `configure_zinc_scala`

Commit:
Introduce a use_ijar ScalaConfiguration value to support Scala 3 and 2

Using ijar on Scala 3 jars currently breaks things. Using ijar on Scala 2
does work. In order to support both of these things simultaneously, I
introduced a config value on the Scala compiler, which informs whether the
ijar or non-ijar version of the target should be used.

I changed the Scala 3 test to make sure we had Scala 3 depending on Scala 2
and vice versa.

Ultimately it would be nice to get rid of this configuration value by
fixing ijars for Scala 3.

SrodriguezO commented on tests/scala/scala3/BUILD:
```diff
@@ -40,5 +40,5 @@ scala_library(
     name = "example-4",
     srcs = ["Example4.scala"],
-    scala = "//scala:3",
+    scala = "//scala:2_13",
```
How come you change this?

jjudd replied:
So I have both scala 2 code depending on 3 and scala 3 depending on 2 in this test.

SrodriguezO reviewed (approved):
one last comment, but LGTM :)

jjudd commented:
So apparently GitHub actions runners don't seem to be cleaning up after themselves and there is some kind of weird MacOS C++ toolchain issue after squashing. I'm going to merge and if it shows up on master, then we can deal with it then.

This does suggest we might need to do something with the GHA build if it leaves state in between builds.
